src/InsetOrderOptimizer.cpp

```diff
@@ -329,5 +263,55 @@ std::unordered_set<std::pair<const ExtrusionLine*, const ExtrusionLine*>> InsetO
     return order;
 }
 
+constexpr bool InsetOrderOptimizer::shouldReversePath(const bool use_one_extruder, const bool current_extruder_is_wall_x, const bool outer_to_inner)
+{
+    if (use_one_extruder && current_extruder_is_wall_x)
+    {
+        return ! outer_to_inner;
+    }
+    return current_extruder_is_wall_x;
+}
 
+std::vector<ExtrusionLine> InsetOrderOptimizer::getWallsToBeAdded(const bool reverse, const bool use_one_extruder)
+{
+    ranges::any_view<VariableWidthLines> view;
+    if (reverse)
+    {
+        if (use_one_extruder)
+        {
+            view = paths | ranges::views::reverse;
+        }
+        else
+        {
+            view = paths | ranges::views::reverse | ranges::views::drop_last(1);
+        }
+    }
+    else
+    {
+        if (use_one_extruder)
+        {
+            view = paths | ranges::views::all;
+        }
+        else
+        {
+            view = paths | ranges::views::take_exactly(1);
+        }
+    }
+    return view | ranges::views::join | ranges::views::remove_if(ranges::empty) | ranges::to_vector;
+}
+
+void InsetOrderOptimizer::dfs(auto node, isGraph auto& dag, isSet auto& visited)
+{
+    if (ranges::contains(visited, node))
```

Comment:
@casperlamboo if you decided on using a "shadow" `unordered_set` for the visited, you should also use the member function `find()` instead of the ranges. Which is much slower for unordered maps and sets. In order to keep this also working for Vectors, you would then also need to update the concept `isSet` and create use an `if constexpr(std::is_same<decltype(visited), ....>)` to create the different execution branches for the contains logic

Reply:
same goes for the `push_back` I think.